src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantClient.cs

```diff
@@ -0,0 +1,350 @@
+    private readonly string? _threadId;
+
+    /// <summary>Initializes a new instance of the <see cref="OpenAIAssistantClient"/> class for the specified <see cref="AssistantClient"/>.</summary>
+    /// <param name="assistantClient">The underlying client.</param>
+    /// <param name="assistantId">The ID of the assistant to use.</param>
+    /// <param name="threadId">
+    /// The ID of the thread to use. If not supplied here, it should be supplied per request in <see cref="ChatOptions.ChatThreadId"/>.
+    /// If none is supplied, a new thread will be created for a request.
+    /// </param>
+    public OpenAIAssistantClient(AssistantClient assistantClient, string assistantId, string? threadId)
+    {
+        _assistantClient = Throw.IfNull(assistantClient);
+        _assistantId = Throw.IfNull(assistantId);
+        _threadId = threadId;
+
+        Metadata = new("openai");
+    }
+
+    /// <inheritdoc />
+    public ChatClientMetadata Metadata { get; }
+
+    /// <inheritdoc />
+    public object? GetService(Type serviceType, object? serviceKey = null)
+    {
+        _ = Throw.IfNull(serviceType);
+
+        return
+            serviceKey is not null ? null :
+            serviceType == typeof(AssistantClient) ? _assistantClient :
+            serviceType.IsInstanceOfType(this) ? this :
+            null;
+    }
+
+    /// <inheritdoc />
+    public Task<ChatCompletion> CompleteAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, CancellationToken cancellationToken = default) =>
+        CompleteStreamingAsync(chatMessages, options, cancellationToken).ToChatCompletionAsync(coalesceContent: true, cancellationToken);
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        // Extract necessary state from chatMessages and options.
+        (RunCreationOptions runOptions, List<FunctionResultContent>? toolResults) = CreateRunOptions(chatMessages, options);
+
+        // Get the thread ID.
+        string? threadId = options?.ChatThreadId ?? _threadId;
+        if (threadId is null && toolResults is not null)
+        {
+            Throw.ArgumentException(nameof(chatMessages), "No thread ID was provided, but chat messages includes tool results.");
+        }
+
+        // Get the updates to process from the assistant. If we have any tool results, this means submitting those and ignoring
+        // our runOptions. Otherwise, create a run, and a thread if we don't have one.
+        IAsyncEnumerable<StreamingUpdate> updates;
+        if (GetRunId(toolResults, out List<ToolOutput>? toolOutputs) is string existingRunId)
+        {
+            updates = _assistantClient.SubmitToolOutputsToRunStreamingAsync(threadId, existingRunId, toolOutputs, cancellationToken);
```

Comment:
Should we also throw if, at this point, we see there are any message contents *other* than FRCs?

Am I right to think that if there are, that additional content would just get lost at the point? (Not 100% certain if I'm visualising the flow correctly)

Reply:
> Am I right to think that if there are, that additional content would just get lost at this point?

Yes

> Should we also throw if, at this point, we see there are any message contents other than FRCs?

I'm a little hesitant to do that. In general we've been operating under a scheme where unrecognized content / options are just ignored rather than producing an error.